libtransmission/settings.cc

```diff
@@ -23,122 +23,64 @@
-        { "critical", TR_LOG_CRITICAL },
-        { "debug", TR_LOG_DEBUG },
-        { "error", TR_LOG_ERROR },
-        { "info", TR_LOG_INFO },
-        { "off", TR_LOG_OFF },
-        { "trace", TR_LOG_TRACE },
-        { "warn", TR_LOG_WARN },
-    } };
-
-    static auto constexpr PreallocationKeys = Lookup<tr_open_files::Preallocation, 5U>{ {
-        { "off", tr_open_files::Preallocation::None },
-        { "none", tr_open_files::Preallocation::None },
-        { "fast", tr_open_files::Preallocation::Sparse },
-        { "sparse", tr_open_files::Preallocation::Sparse },
-        { "full", tr_open_files::Preallocation::Full },
-    } };
-
-    static auto constexpr VerifyModeKeys = Lookup<tr_verify_added_mode, 2U>{ {
-        { "fast", TR_VERIFY_ADDED_FAST },
-        { "full", TR_VERIFY_ADDED_FULL },
-    } };
-
-    static auto constexpr PreferredTransportKeys = Lookup<tr_preferred_transport, TR_NUM_PREFERRED_TRANSPORT>{ {
-        { "utp", TR_PREFER_UTP },
-        { "tcp", TR_PREFER_TCP },
-    } };
-};
-
-template<>
-std::optional<bool> VariantConverter::load(tr_variant const& src)
-{
-    if (auto val = src.get_if<bool>(); val != nullptr)
-    {
-        return *val;
-    }
-
-    return {};
-}
-
-template<>
-tr_variant VariantConverter::save(bool const& val)
-{
-    return val;
-}
+template<typename T, size_t N>
+using Lookup = std::array<std::pair<std::string_view, T>, N>;
 
 // ---
 
-template<>
-std::optional<std::chrono::milliseconds> VariantConverter::load(tr_variant const& src)
+bool load_bool(tr_variant const& src, bool* tgt)
 {
-    if (auto val = src.get_if<int64_t>(); val != nullptr)
+    if (auto val = src.get_if<bool>(); val != nullptr)
     {
-        return std::chrono::milliseconds(*val);
+        *tgt = *val;
+        return true;
     }
 
-    return {};
+    return false;
 }
 
-template<>
-tr_variant VariantConverter::save(std::chrono::milliseconds const& val)
+tr_variant save_bool(bool const& val)
 {
-    return val.count();
+    return val;
 }
 
 // ---
 
-template<>
-std::optional<double> VariantConverter::load(tr_variant const& src)
+bool load_double(tr_variant const& src, double* tgt)
 {
     if (auto val = src.get_if<double>(); val != nullptr)
     {
-        return *val;
+        *tgt = *val;
+        return true;
     }
 
-    return {};
+    return false;
 }
 
-template<>
-tr_variant VariantConverter::save(double const& val)
+tr_variant save_double(double const& val)
 {
     return val;
 }
 
 // ---
 
-template<>
-std::optional<tr_encryption_mode> VariantConverter::load(tr_variant const& src)
+auto constexpr EncryptionKeys = Lookup<tr_encryption_mode, 3U>{ {
+    { "required", TR_ENCRYPTION_REQUIRED },
+    { "preferred", TR_ENCRYPTION_PREFERRED },
+    { "allowed", TR_CLEAR_PREFERRED },
+} };
+
+bool load_encryption_mode(tr_variant const& src, tr_encryption_mode* tgt)
 {
-    static constexpr auto Keys = EncryptionKeys;
+    constexpr auto const& Keys = EncryptionKeys;
```

Comment:
*(nit)* Only this particular load function is changed to use a const reference, better keep it consistent across the board.

Reply:
yep, you're right. That `const` is even generating a new warning.

Fixed 30603b4485cb60dab157e572b29ddcf8481ac2f0